### ISSUE TITLE
Neo4jCsvPublisher Speed Optimization (Parallelism) 

### ISSUE THREAD

**Issue by chonyy** [amundsen-io/amundsen#1610]
Hi Team,
I’m wondering if there’s a plan to apply **multiprocessing** on the publishers. We have a large amount of metadata in our production, which ended up running 3 million queries on `neo4j` . It takes about 90 minutes to finish.

To investigate the bottleneck, I looked into the code and logged the time elapsed for each step in a single iteration in the `_publish_node` [function](https://github.com/amundsen-io/amundsen/blob/699652f7e9a884a05000a9ca00bce72179cfa593/databuilder/databuilder/publisher/neo4j_csv_publisher.py#L253). This is the result

- Neo4j query: 0.1ms
- Create statement: 1ms
- Others: super fast, doesn’t matter
![image](https://user-images.githubusercontent.com/44046550/144763323-dcdecaad-8cb6-4654-8fad-4a3ffb56102e.png)

Surprisingly, the **bottleneck** is not the db query, it’s the statement creation. The process is basically

1. loop each row in csv
2. parse the row into a dictionary
3. loop through each key value pair in the dictionary to get the props
4. fill the statement `Jinjia`  template with the props
5. execute the query with the statement

I’m thinking that instead of read a row => create a node in graph db one by one, maybe we could use multiprocessing to speed up the process. I believe there will be no dependency issue as long as we publish all the nodes before publishing relations, which is already handled in the current codebase.
I’m planning on implementing multiprocessing for this, is there any potential problem? Like dependency, graph db load, etc..

## Expected Behavior or Use Case
Speed up the performance of the publisher. Currently, a 90 min sync is not acceptable for our use case 😢 

## Service or Ingestion ETL
Ingestion ETL, publisher

## Possible Implementation

Thanks to @dkunitsk 's idea, I think there are three possible implementations

1. Multiprocessing on call side
2. Multiprocessing on Neo4j publisher
3. Neo4j UNWIND (Batch processing)

![image](https://user-images.githubusercontent.com/44046550/144763755-5aaeb2f4-4315-4cac-9326-a3a3eb13d859.png)
![image](https://user-images.githubusercontent.com/44046550/144763760-5e92731a-5fc4-4712-9ac1-5bc79750fcda.png)
![image](https://user-images.githubusercontent.com/44046550/144763761-53c3fcae-cfa2-4993-b0b6-8d55fc54e89d.png)

```python
class HiveParallelIndexer:
    # Shim for adding all node labels to the NEO4J_DEADLOCK_NODE_LABELS config
    # which enables retries for those node labels. This is important for parallel writing
    # since we see intermittent Neo4j deadlock errors relatively often.
    class ContainsAllList(list):
        def __contains__(self, item):
            return True

    def __init__(self, publish_tag: str, parallelism: int):
        self.publish_tag = publish_tag
        self.parallelism = parallelism

    def __call__(self, worker_index: int):
        # Sharding:
        #   - take the md5 hash of the schema.table_name
        #   - convert the first 3 characters of the hash to decimal (3 chosen arbitrarily)
        #   - mod by total number of processes
        where_clause_suffix = """
            WHERE MOD(CONV(LEFT(MD5(CONCAT(d.NAME, '.', t.TBL_NAME)), 3), 16, 10), {total_parallelism}) = {worker_index}
            AND t.TBL_TYPE IN ('EXTERNAL_TABLE', 'MANAGED_TABLE', 'VIRTUAL_VIEW')
            AND (t.VIEW_EXPANDED_TEXT != '/* Presto View */' OR t.VIEW_EXPANDED_TEXT is NULL)
        """.format(total_parallelism=self.parallelism,
            worker_index=worker_index)

        # configs relevant for multiprocessing
        job_config = ConfigFactory.from_dict({
            'extractor.hive_table_metadata.{}'.format(HiveTableMetadataExtractor.WHERE_CLAUSE_SUFFIX_KEY):
                where_clause_suffix,
            # keeping this relatively low, in our experience, reduces neo4j deadlocks
            'publisher.neo4j.{}'.format(neo4j_csv_publisher.NEO4J_TRANSACTION_SIZE):
                100,
            'publisher.neo4j.{}'.format(neo4j_csv_publisher.NEO4J_DEADLOCK_NODE_LABELS):
                HiveParallelIndexer.ContainsAllList(),
        })
        job = DefaultJob(conf=job_config,
                         task=DefaultTask(
                             extractor=HiveTableMetadataExtractor(),
                             loader=FsNeo4jCSVLoader()),
                         publisher=Neo4jCsvPublisher())
        job.launch()


parallelism = 16
indexer = HiveParallelIndexer(
    publish_tag='2021-12-03'
    parallelism=parallelism)

with multiprocessing.Pool(processes=parallelism) as pool:
    def callback(_):
        # fast fail in case of exception in any process
        print('terminating due to exception')
        pool.terminate()
    res = pool.map_async(indexer, [i for i in range(parallelism)], error_callback=callback)
    res.get()
```

## Screenshots of Slack Discussion

![image](https://user-images.githubusercontent.com/44046550/144763614-23eca0fe-b78a-4cc0-bbb2-141181416e44.png)
![image](https://user-images.githubusercontent.com/44046550/144763544-64c1c4f5-f5be-44ab-80cc-c736e4cf497a.png)

**Comment by chonyy** [comment]
The `where_clause_suffix` provided in the sample code above is for MySQL. The syntax is not compatible for PostgreSQL. For Postgres it should be replaced to 
```python
where_clause_suffix = """
            WHERE MOD(('x' || lpad(LEFT(MD5(CONCAT(d."NAME", '.', t."TBL_NAME")), 3), 8, '0'))::bit(32)::int, {total_parallelism}) = {worker_index}
            AND t."TBL_TYPE" IN ('EXTERNAL_TABLE', 'MANAGED_TABLE', 'VIRTUAL_VIEW')
            AND (t."VIEW_EXPANDED_TEXT" != '/* Presto View */' OR t."VIEW_EXPANDED_TEXT" is NULL)
        """.format(total_parallelism=self.parallelism, worker_index=worker_index)
```

**Comment by chonyy** [comment]
When working with Airflow pipelines, it would be better to split it into multiple tasks in a dag instead of running multiprocess in a single task. The sample code is provided below
```python


# Extract metastore table/column metadata from metastore and publish it to neo4j
def run_multi_hive_metastore_job(parallelism, worker_index):
    # Sharding:
    #   - take the md5 hash of the schema.table_name
    #   - convert the first 3 characters of the hash to decimal (3 chosen arbitrarily)
    #   - mod by total number of processes
    where_clause_suffix = """
        WHERE MOD(('x' || lpad(LEFT(MD5(CONCAT(d."NAME", '.', t."TBL_NAME")), 3), 8, '0'))::bit(32)::int, {total_parallelism}) = {worker_index}
        AND t."TBL_TYPE" IN ('EXTERNAL_TABLE', 'MANAGED_TABLE', 'VIRTUAL_VIEW')
        AND (t."VIEW_EXPANDED_TEXT" != '/* Presto View */' OR t."VIEW_EXPANDED_TEXT" is NULL)
    """.format(total_parallelism=parallelism, worker_index=worker_index)

    tmp_folder = '/var/tmp/amundsen/table_metadata' + str(worker_index)
    node_files_folder = f'{tmp_folder}/nodes/'
    relationship_files_folder = f'{tmp_folder}/relationships/'

    job_config = ConfigFactory.from_dict({
        'extractor.hive_table_metadata.{}'.format(HiveTableMetadataExtractor.WHERE_CLAUSE_SUFFIX_KEY): where_clause_suffix,
        # keeping this relatively low, in our experience, reduces neo4j deadlocks
        'publisher.neo4j.{}'.format(neo4j_csv_publisher.NEO4J_TRANSACTION_SIZE): 300,
        f'extractor.hive_table_metadata.extractor.sqlalchemy.{SQLAlchemyExtractor.CONN_STRING}': metastore_postgresql_connection_string(),
        f'loader.filesystem_csv_neo4j.{FsNeo4jCSVLoader.NODE_DIR_PATH}': node_files_folder,
        f'loader.filesystem_csv_neo4j.{FsNeo4jCSVLoader.RELATION_DIR_PATH}': relationship_files_folder,
        f'publisher.neo4j.{neo4j_csv_publisher.NODE_FILES_DIR}': node_files_folder,
        f'publisher.neo4j.{neo4j_csv_publisher.RELATION_FILES_DIR}': relationship_files_folder,
        f'publisher.neo4j.{neo4j_csv_publisher.NEO4J_END_POINT_KEY}': neo4j_endpoint,
        f'publisher.neo4j.{neo4j_csv_publisher.NEO4J_USER}': neo4j_user,
        f'publisher.neo4j.{neo4j_csv_publisher.NEO4J_PASSWORD}': neo4j_password,
        f'publisher.neo4j.{neo4j_csv_publisher.JOB_PUBLISH_TAG}': timestamp,  # give the timestamp to each node in neo4j
    })
    job = DefaultJob(conf=job_config,
                        task=DefaultTask(
                            extractor=HiveTableMetadataExtractor(),
                            loader=FsNeo4jCSVLoader()),
                        publisher=Neo4jCsvPublisher())
    job.launch()

for idx in range(parallelism):
    hive_metastore_task = PythonOperator(
        task_id='hive_metastore_task' + str(idx),
        python_callable=run_multi_hive_metastore_job,
        dag=dag,
        op_kwargs={'parallelism': parallelism, 'worker_index': idx})
    kick_off_dummy >> hive_metastore_task >> trino_view_task >> staleness_removal_task >> publish_elastic_search_task

```

**Comment by chonyy** [comment]
Based on my experience, the ingestion speed is about 3 times faster by applying multiprocessing on the caller side. Since the performance improvement will be bounded by the neo4j instance, it wouldn't help much if we also implement multiprocessing on the publisher side. This means that the only other possible optimization that might help is neo4j `UNWIND`.

**Comment by mgorsk1** [comment]
Thanks @chonyy and @dkunitsk for bringing this up, I've stumbled upon this as well recently and have following thoughs:

1. I think looking into parallelization on caller (executor) side is not scalable as it'd require rework on every extractor. The benefit, however, would be that it's automatically available for any publisher.
2. I am totally in favor of improving neo4j publisher as it would support every extractor. I took initial spin with using current code with multiprocessing, however it still results in some locks. The improvement was 2.5 times faster with 15k tables (500k columns). The code is here https://github.com/amundsen-io/amundsen/commit/40791ad836d4bb1742a82b5388e77fb3e7f3d8a9 - if you have any hints on how to avoid locks it could be short-term solution
3. long-term unwind seems like a way to go, we already use bulk loads in ES publisher and should introduce it for neo4j too.

**Comment by mgorsk1** [comment]
Thanks @chonyy and @dkunitsk for bringing this up, I've stumbled upon this as well recently and have following thoughs:

1. I think looking into parallelization on caller (executor) side is not scalable as it'd require rework on every extractor. The benefit, however, would be that it's automatically available for any publisher.
2. I am totally in favor of improving neo4j publisher as it would support every extractor. I took initial spin with using current code with multiprocessing, however it still results in some locks. The improvement was 2.5 times faster with 15k tables (500k columns). The code is here https://github.com/amundsen-io/amundsen/commit/40791ad836d4bb1742a82b5388e77fb3e7f3d8a9 - if you have any hints on how to avoid locks it could be short-term solution
3. long-term unwind seems like a way to go, we already use bulk loads in ES publisher and should introduce it for neo4j too.

**Comment by zacr** [comment]
@mgorsk1, @chonyy, @dkunitsk I created a version of the publisher that uses UNWIND and APOC library. It's 5x faster for me on my test dataset (1500 tables, 650k columns). Still rough around the edges and I did it based on the CsvTable*Extractors so not sure if I baked in any assumptions that won't hold with other extractors. It will need some testing and I likely introduced some issues - but workable updated cypher queries should be in place and planning on running it for a few weeks to shake it out. There is additional room for optimization possibly; not doing anything in parallel to avoid locks though apoc.iterate supports parallelism and retries.  If anybody wants to play with it and try to get it over the line happy to toss you the code.

**Comment by chonyy** [comment]
Hi @zacr , thanks for looking into this!

I'm not super familiar with the capabilities of `UNWIND`, but the 5x speedup sounds promising 🚀
One thing that I would like to clarify is currently there's no any kind of parallelism in your version right? Does this mean that we could have at least 10x speedup if we combined both multiprocessing and `UNWIND`?

Also, I'm interested in your implementation, could you point me to the commit or the branch? Thanks in advance 🙏

**Comment by zacr** [comment]
Wasn't sure if I am allowed to make a branch in the main repo or if I am supposed to fork first. Hopefully you can see this, I did it as a new file so could run both side by side: https://github.com/SaltIO/amundsen/blob/neo4j_csv_publisher_apoc/databuilder/databuilder/publisher/neo4j_csv_publisher_apoc.py

I didn't do any query optimization or tuning, the goal was to just try to get functional cypher queries using [apoc.periodic.iterate](https://neo4j.com/labs/apoc/4.2/overview/apoc.periodic/apoc.periodic.iterate/) which has support for parallelization. I suspect there is some meat on the bone by investigating and tuning these new cypher queries. The most recent version looks more like 3.5x faster instead of 5x, but I didn't do any serious timing. If you turn on Parallel:true and retries: 1 it goes 5x but locks will cause you to drop maybe 0.3% of the inserts. These cypher queries may be on the path to 7x+ with tuning unless I did them wrong.

After you investigate if you think this APOC based solution has promise and you or someone else wants to run with it that is great by me. In the meantime I am just going to try to let it run in a few projects with different extractors over the next few weeks to surface issues.

**Comment by afuzzyriff** [comment]
Hello - I stumbled across this issue as I was facing a similar problem of poor performance with large Neo4j writes.

@zacr  Thank you for sharing your Neo4j CSV Publisher with APOC code. I reviewed the class and started to test this in my team's Amundsen POC implementation for Neo4j CSV publishing. The custom class provided seems to be working nicely.

I have tested a batch of data which resulted in writing 4,848,906 nodes and 9,715,678 relationships in 5704.64 seconds (~95 minutes) using an 8k batch size. I don't have good non-APOC query timing results as that scenario was running rather slowly, but I do know this is a vast timing improvement.

**Comment by zacr** [comment]
Great to hear! Our testing has been going well - we have been using this class for all publishes the last month and haven't seen any issues.  I was hoping to find time to try a Neo4J version upgrade and do a little testing on that. I think the INDEX creation code in that publisher would need to be tweaked for a newer version of Neo4J. Was also thinking about looking at the parallelism one more time to see if there is any speed up there. Lots of good info in this article to try out still: https://neo4j.com/blog/nodes-2019-best-practices-to-make-large-updates-in-neo4j/

**Comment by akiratwang** [comment]
> 

Regarding this, would be a simple merge into the main repo or can I download it and add it to mine. I presume the change would be as simple as replacing `neo4j_csv_publisher` with `neo4j_csv_publisher_apoc`?

There were also some issues with getting APOC installed alongside neo4j so if anyone has issues, let me know as I have been able to resolve it.

**Comment by akiratwang** [comment]
Looking to deploy this in prod and will let you know how it goes. For starters without APOC:
- Roughly 11k tables and 230k columns exposed to front-end users
- ~3.9 million nodes to commit
- Original runtime of 1.5-2.5 hours

Will update once deployment is completed!

**Comment by akiratwang** [comment]
With APOC stats:
- About the same number of tables/columns/nodes
- Runtime of just under 25 minutes...

Definitely an improvement!!! If you want to open a PR with this, I can add the relevant arguments to the docker compose config to have `apoc` installed to the corresponding version of `neo4j`. See https://github.com/amundsen-io/amundsen/discussions/1801

**Comment by akiratwang** [comment]
For anyone else looking to implement @zacr 's version here: https://github.com/SaltIO/amundsen/blob/neo4j_csv_publisher_apoc/databuilder/databuilder/publisher/neo4j_csv_publisher_apoc.py

1. Add the script to `amundsendatabuilder/publisher/`
2. Open up your loader script. (As an example, you can open `amundsendatabuilder/example/scripts/sample_postgres_loader.py`) 
3. Replace Imports as follows
```python
# Original Below
# from databuilder.publisher import neo4j_csv_publisher
# from databuilder.publisher.neo4j_csv_publisher import Neo4jCsvPublisher

# New imports 
from databuilder.publisher import neo4j_csv_publisher_apoc
from databuilder.publisher.neo4j_csv_publisher_apoc import Neo4jCsvPublisherApoc
```
4. In your `job_config` dict, replace it with the new imports. For example,
```
# Original config line
# f'publisher.neo4j.{neo4j_csv_publisher.NEO4J_TRANSACTION_SIZE}'

# Becomes to this
f'publisher.neo4j.{neo4j_csv_publisher_apoc.NEO4J_TRANSACTION_SIZE}'
```

Many thanks again for building this script!

**Comment by zacr** [comment]
Awesome @akiratwang, so glad it worked well. I opened a PR here: https://github.com/amundsen-io/amundsen/pull/1836. If you want to add anything to it go for it.

**Comment by akiratwang** [comment]
Awesome stuff, I've found a small issue I can fix w.r.t `CREATE_ONLY_NODES` so I'll add the changes to the PR by early next week!

**Comment by akiratwang** [comment]
Looks solid! I've added a couple comments to mitigate some niche issues when specifying create-only nodes. Was wondering how your team got APOC working with neo4j without specifying it in the docker compose scripts?

**Comment by chonyy** [comment]
@akiratwang @zacr Thanks for all the progress here! I'm so glad to see this issue finally turned into a PR and excited to see it get into production. The speed improvement is just amazing 🚀

**Comment by akiratwang** [comment]
> Oh, I just checked and we do have it in the docker-compose. We have this in the neo4j section: volumes: - ./configs/neo4j/conf:/var/lib/neo4j/conf - ./configs/neo4j/plugins:/var/lib/neo4j/plugins
> 
> And then in the plugins folder we have apoc-3.5.0.17-all.jar.
> 
> We didn't modify the docker-compose in the Amundsen repo, we have a separate repo for deployment based on [Stemma's original recommendation](https://www.stemma.ai/blog-post/amundsen-deployment-best-practices). It works great. I didn't even think about going back and adding it into main repo but you are right, thats probably a good idea. I will hop over and check out your PR comments this weekend and try to knock them out. Thank you for pushing this over the line!!!!

Ahh yep that would make sense, was going through the salt-io repo and couldn't find it. Many thanks!

**Comment by chonyy** [comment]
@zacr Thanks a lot for your amazing work! I couldn't wait for your PR to get merged. I tested your code in PR and wanted to share the exciting experiment results with you!

### Execution time before apoc apoc: 100 ~ 120 minutes
<img width="214" alt="image" src="https://user-images.githubusercontent.com/44046550/170977027-83f6f101-420b-4642-9bc4-9ceb5869dc08.png">

### Execution time after apoc apoc: 7 ~ 11 minutes
<img width="219" alt="image" src="https://user-images.githubusercontent.com/44046550/170977065-9b6cbb65-808c-4a15-8856-ea38af6eb3ba.png">

Originally, without any optimization, it will have to take **5 hours** for each sync in our environment.
- Original: 5 hours
- After doing [caller side partition](https://github.com/amundsen-io/amundsen/issues/1610#issuecomment-1000019662): 5 hours -> 2 hours
- After applying `Neo4jCsvPublisherApoc`: 120min -> 10min

This is like an average 30x improvement, which is pretty insane IMO 🚀

**Comment by chonyy** [comment]
The PR related to this is here: https://github.com/amundsen-io/amundsen/pull/1877
